### ISSUE TITLE
unsupported atomic accessors ?

### ISSUE THREAD

**Issue by fwyzard** [illuhad/hipSYCL#71]
This code snippet:
```c++
  // construct a SYCL queue for the selected device
  auto queue = cl::sycl::queue(device);
  
  // SYCL buffers
  int counter = 0;
  auto counter_buf = cl::sycl::buffer<int>(& counter, 1);
  
  // submit the kernel to the queue
  queue.submit([&](cl::sycl::handler &cgh) {
    // access the SYCL buffers from the device kernel
    auto counter_d = counter_buf.get_access<cl::sycl::access::mode::atomic>(cgh);
    
    // launch the kernel
    cgh.parallel_for_work_group<class count_groups>(
        cl::sycl::range<1>{num_work_groups},
        cl::sycl::range<1>{work_group_size},
        [=](cl::sycl::group<1> group) {
        
      // print the id of all the groups
      printf("group id: %lu\n", group.get_id(0));
      
      // print the id of all the threads
      group.parallel_for_work_item([&](cl::sycl::h_item<1> item) {
        printf("global thread id: %zu\n", item.get_global_id(0));
      });
      
      cl::sycl::atomic_fetch_add(counter_d[0], 1);
    });
```

it fails with

```
test.cc:53:43: error: no viable overloaded operator[] for type 'const cl::sycl::accessor<int, 1, cl::sycl::access::mode::atomic, cl::sycl::access::target::global_buffer,
      cl::sycl::access::placeholder::false_t>'
      cl::sycl::atomic_fetch_add(counter_d[0], 1);
                                 ~~~~~~~~~^~
```

and in fact the atomic accessors in [include/CL/sycl/accessor.hpp](https://github.com/illuhad/hipSYCL/tree/master/include/CL/sycl/accessor.hpp) are commented out:
```c++
  /* Available only when: accessMode == access::mode::atomic && dimensions == 0*/
  //operator atomic<dataT, access::address_space::global_space> () const;

  /* Available only when: accessMode == access::mode::atomic && dimensions > 0*/
  //atomic<dataT, access::address_space::global_space> operator[](
  //    id<dimensions> index) const;

  //atomic<dataT, access::address_space::global_space> operator[](
  //    size_t index) const;
```

What is the reason for that ?

**Comment by illuhad** [comment]
The `atomic<>` class template is still incomplete, so we haven't yet enabled support for atomic accessors. However, at the moment I believe the only functions missing from `atomic<>` are now `load()` and `store()`, so we could actually start enabling these accessors. Would you like to file a PR for that?

**Comment by illuhad** [comment]
Regarding your code snippet: Note that are for parallel hierarchical for invocations, we do not yet make sure that statements inside `parallel_for_work_group` like your `printf` are executed only once per work group, as defined by the spec. Instead you will see these statements being executed for each work item. Implementing this is a bit more complicated, so we haven't done it yet.
Variable declarations in `parallel_for_work_group` are supported however and will end up in local memory.

**Comment by fwyzard** [comment]
> Instead you will see these statements being executed for each work item. Implementing this is a bit more complicated, so we haven't done it yet.

I asse... FWIW, that is also the behaviour I am seeing with the latest ComputeCpp from CodePlay :-(

**Comment by illuhad** [comment]
Interesting! I'm pretty surprised by that. After all, ComputeCpp should be SYCL-conformant... I'm going to have to check the spec again to make sure I didn't misinterpret anything.

**Comment by fwyzard** [comment]
I just received confirmation that it is a bug: https://support.codeplay.com/t/incorrect-behaviour-of-parallel-for-work-group/173 .

**Comment by illuhad** [comment]
I see, thanks for the update!

**Comment by reguly** [comment]
Any idea when atomics may get supported? I have several interesting use cases where I am getting good benefits in CUDA with the use of atomics. Unfortunately I know far too little about compilers to add the support myself...

**Comment by illuhad** [comment]
No need to know about compilers, atomics are implemented purely as part of the hipSYCL device library :smiley: 
Could you please tell me a little more about the functions you need? Do you really need atomic accessors, or are atomic operations sufficient? For the latter, you should already be able to use the `atomic` class directly as implemented [here](https://github.com/illuhad/hipSYCL/blob/master/include/CL/sycl/atomic.hpp). If you have a pointer to device memory, you can just create an `atomic` object for that pointer and start using atomic operations. The only major unimplemented features at the moment are raw `store()` and `load()` operations.
If you truly need the atomic accessor interface - I can have a look at that as one of my next todos. At the moment, I'm working on CPU/OpenMP performance optimizations. When I'm done with that I can see if I can find time to check on atomics.

**Comment by reguly** [comment]
I only need fetch_add on various global memory positions, so I don't think I need the atomic accessors. Pardon my ignorance, but it is not clear to me how I actually use this...

I have the following:
`cl::sycl::buffer<double,1> *arg6_buffer = ...`
...
`auto ind_arg3 = (*arg6_buffer).template get_access<cl::sycl::access::mode::read_write>(cgh);`
...
`cl::sycl::atomic<double> a(ind_arg3[0+map2idx*4]); a.fetch_add(arg6_l[0]);`

but it's complaining about the template parameter to atomic:

> /usr/local/bin/../include/CL/sycl/atomic.hpp:53:3: note: candidate template ignored: could not match 'multi_ptr<type-parameter-0-0, cl::sycl::access::address_space::global_space>' against 'double'

**Comment by illuhad** [comment]
Sorry for not giving an example right away - I suppose it can be a bit counterintuitive if you aren't familiar with the way the SYCL API encapsulates pointers. Haven't explicitly tested it, but try something like the following:
```cpp
cl::sycl::atomic<double> my_atomic{
  cl::sycl::global_ptr<double>{
     &(ind_arg3[0+map2idx*4) /* double pointer goes here */
  }
};
```
The atomic accessor API provides essentially a shortcut to that. `global_ptr<T>` is a typedef to `multi_ptr<T, sycl::access::address_space::global_space>`. I think the main issue of your attempt is that `ind_arg3[0+map2idx*4]` is not really a pointer, but a `double` (or double reference). It may perhaps already work by replacing only that with `&(ind_arg3[0+map2idx*4)` since pointers should be implicitly convertible to `multi_ptr`.

**Comment by reguly** [comment]
Thank you - I did get this working. However, I had to make some changes in the header files:

- fetch_add does not accept floats and doubles currently - for my use case I just commented out the enable_if_t
- the atomic object is marked volatile, and so is the multi_ptr held within it, however, the get() member function of multi_ptr is not marked volatile, so that gave me errors - marking it volatile solved the issue (I am not sure though if this has negative effects elsewhere).

Re: pointers to buffers - I am integrating SYCL into the OP2 library (https://github.com/OP-DSL/OP2-Common/tree/feature/sycl) which has a C API, and so to get through it, I need typecasting down to plain pointers, and then back. Both the back-end, which manages memory, and the auto-generated code is C++, but the interface between them is C...

**Comment by illuhad** [comment]
Thanks for the feedback! 
* The current SYCL spec explicitly does explicitly only allow integer atomic additions. In that sense, the behavior of hipSYCL is correct. Of course, I can see the value of floating point atomics. It may be useful to implement a hipSYCL extension here. HIP supports float atomic additions, but not `double`. We could however add a special code path for double atomic support when compiling for NVIDIA since CUDA supports that.
Unfortunately, for portable SYCL code, the only solution for floating point atomics at the moment is an atomic CAS loop, which would probably not be efficient.
* Regarding `volatile`: I don't think `get()` should be volatile, since `multi_ptr` is the generic SYCL pointer class and may be used widely in applications - using volatile here may slow things down for some applications. I'll see if I can come up with a fix inside the `atomic` class.

---
Ah, so buffer pointers are caused by the devil of compatibility with legacy C interfaces. I see :)

**Comment by illuhad** [comment]
@reguly, please check out PR #88 - it should fix both issues when `#define HIPSYCL_EXT_FP_ATOMICS` is inserted before including `CL/sycl.hpp`.

**Comment by Alkeryn** [comment]
Just wanted to point out that fetch_add(float) does not work even tho float are part of the sycl spec
![image](https://user-images.githubusercontent.com/11599075/70953831-096dc600-206b-11ea-9c1f-75d7f63a3c0f.png)

**Comment by Alkeryn** [comment]
just wanted to point out that some types of the sycl specs don't work with atomic
![image](https://user-images.githubusercontent.com/11599075/70953942-8a2cc200-206b-11ea-84d3-3bbfd6523a8f.png)
ex: foo.fetch_add(long);

**Comment by illuhad** [comment]
`atomic` operators implemented in PR #210
